Studio-Lovelies/GG-JointJustice-Unity
# #19 - Audio Controller

## Summary


Originally this was meant to be just the "Music Player" side of the Audio Controller, but playing sound effects was pretty trivial by comparison so I included that too.

#### Music
Music is played as a single audio source attached directly to the Controller's GameObject, when you play a new track it fades out (over 1 second), then switches track, then fades in (also over 1 second). This does not (yet) support crossfade (fade out A while fading in B). Although I could make some adjustments to get us there if the sound team wants it.

There's also a "SettingsVolume" which is currently a field on AudioController but should eventually come from some core "Settings" object. This scales the total volume tween for the fader, so the fade always takes 1 second, but the amount that it fades in that time will vary depending on the settings volume.

#### Sound Effects
~~Sound Effects work slightly differently, since you apparently need an AudioSource (and therefore an entire game object) to play a sound. I create a child object on AudioController for each sound effect played. However, the game objects stick around after the sound is played, and they're reused if the sound is played multiple times.~~

SoundEffects use PlayOneShot on the existing AudioSource

#### Debug Stuff

Currently there's some clearly marked debug logic that starts out on a particular track and changes the track when you press A or B, and plays a sound when you press C. I can remove this before the final submit, but I left it there to make it easy for people to test.

## Before/after screenshots and/or animated gif

![AudioControllerExample](https://user-images.githubusercontent.com/6920005/130090811-e22d06a8-3151-4311-8fa8-955c0b071c98.gif)

## Testing instructions

Load up the `AudioController - Test Scene` scene and press `A` and `B` to switch between two tracks. Move the SettingsMusicVolume slider up and down while the music is playing. Try adjusting it while it's fading too and see if you get the expected result.

## Additional information

- `[ ]` Changes UI
- `[X]` Introduces new feature
- `[ ]` Removes existing feature
- `[X]` Has associated resource:  #19
